compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -30,6 +30,113 @@
 #include "il/Node_inlines.hpp"
 #include "infra/Bit.hpp"
 
+/**
+ * Helper that evaluates a child node.
+ * @param[in] childReg : the register of the child node being looked at
+ * @param[in] trgReg : the target register
+ * @param[in] child : the child node being looked at
+ * @param[in] parent : the parent of the child node being looked at
+ * @param[in] cg : codegenerator
+ */
+static inline void
+evaluateRegister(TR::Register **childReg, TR::Register *trgReg, TR::Node *child, TR::Node *parent, TR::CodeGenerator *cg)
+   {
+   int64_t value;
+	if (child->getOpCode().isLoadConst() && child->getRegister() == NULL)
+      {
+      value = child->getLongInt();
+      //if(constantIsUnsignedImm12(value))
+         //{
+         //TODO: Cover special cases for immediate loads
+         //}
+      //else
+         //{
+         *childReg = cg->allocateRegister();
+         loadConstant64(cg, parent, value, *childReg);
+         //}
+      }
+   else
+      {
+      *childReg = cg->evaluate(child);
+      }
+   if(child->getReferenceCount() == 1 && NULL == trgReg)
+      {
+      trgReg = *childReg;
+      }
+   }
+
+/**
+ * Generic helper that handles a number of similar binary operations.
+ * @param[in] node : calling node
+ * @param[in] regOp : the matching opcode of the caller
+ * @param[in] cg : codegenerator
+ * @return target register
+ */
+static inline TR::Register *
+genericBinaryEvaluator(TR::Node *node, TR::InstOpCode::Mnemonic regOp, TR::CodeGenerator *cg)
+   {
+   int sizeOfOperation = node->getNumChildren();
+   TR::Register *srcReg[sizeOfOperation];
+   TR::Node *firstChild;
+   TR::Node *secondChild;
+   TR::Node *thirdChild;
+   TR::Register *trgReg = NULL;
+   int i;
+
+   if(0 < sizeOfOperation)
+      {
+      firstChild = node->getFirstChild();
+      evaluateRegister(&srcReg[0], trgReg, firstChild, node, cg);
+      }
+   if(1 < sizeOfOperation)
+      {
+      secondChild = node->getSecondChild();
+      evaluateRegister(&srcReg[1], trgReg, secondChild, node, cg);
+      }
+   if(2 < sizeOfOperation)
+      {
+      thirdChild = node->getThirdChild();
+      evaluateRegister(&srcReg[2], trgReg, thirdChild, node, cg);
+      }
+
+   if(NULL == trgReg)
+      {
+      trgReg = cg->allocateRegister();
+      }
+
+   if(1 == sizeOfOperation)
+      {
+      generateTrg1Src1Instruction(cg, regOp, node, trgReg, srcReg[0]);
+      }
+   else if(2 == sizeOfOperation)
+      {
+      generateTrg1Src2Instruction(cg, regOp, node, trgReg, srcReg[0], srcReg[1]);
+      }
+   else
+      {
+      generateTrg1Src3Instruction(cg, regOp, node, trgReg, srcReg[0], srcReg[1], srcReg[2]);
+      }
+
+   if(0 < sizeOfOperation && firstChild->getReferenceCount() > 1)
+      {
+      cg->stopUsingRegister(srcReg[0]);
```

Comment:
Calling `stopUsingRegister` manually should only be done for "scratch registers".  That is, registers that have a limited live range and are not associated with a node (i.e., they don't appear when you ask `node->getRegister()`).  When you decrement the reference count on a node and it determines that its reference count is 1 then it will implicitly call `stopUsingRegister` already.

So, you only want to be calling `stopUsingRegister` here for the scratch registers you created (via `allocateRegister`) that will not live past this evaluator.